02-linked-lists/02-linked-lists-implementation.md

```diff
@@ -0,0 +1,1030 @@
+Still feeling stuck? Check this video walkthrough of the solution.
+<!-- TODO: Add video walkthrough to explanation -->
+
+##### !end-hint
+
+##### !explanation
+An example of a working implementation:
+
+```py
+    def add_last(self, value):
+        if not self.head:
+            self.add_first(value)
+        else:
+            new_node = Node(value)
+            current = self.head
+            while current.next:
+                current = current.next
+            current.next = new_node
+
+```
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: multiple-choice
+* id: 72c16689-c332-4922-bc1b-bf94ee82ce49
+* title: `add_last` Time Complexity
+* points: 1
+
+##### !question
+
+What is the time complexity of `add_last`?
+
+##### !end-question
+
+##### !options
+
+a| O(1)
+b| O(log n)
+c| O(n log n)
+d| O(n)
+e| O(n^2)
+
+##### !end-options
+
+##### !answer
+
+d|
+
+##### !end-answer
+
+
+##### !explanation
+The time complexity will be O(n) or linear as in the worst case, we will loop through every element in our linked list. 
+
+##### !end-explanation
+
+### !end-challenge
+<!-- ======================= END CHALLENGE ======================= -->
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+
+### !challenge
+
+* type: multiple-choice
+* id: c9612963-f091-48cd-b928-d25ed10efa8f
+* title: Space complexity `add_last`
+* points: 1
+
+##### !question
+
+What is the space complexity of `add_last`?
+
+##### !end-question
+
+##### !options
+
+a| O(1)
+b| O(log n)
+c| O(n log n)
+d| O(n)
+e| O(n^2)
+
+##### !end-options
+
+##### !answer
+
+a|
+
+##### !end-answer
+
+
+##### !explanation
+The space complexity will be O(1) or constant as we are not creating any new data structures or call stacks whose size are proportional to the length of our list.
+
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+### Removing a node at a specific index
+
+To remove a node at a specific index, we also have to traverse the list until we find the index before the node to delete. Remove also requires some more complex redirection of pointers. If the node to remove is the head, then the head becomes the next element. In all other cases for a singly linked list, we adjust the prior node's `next` reference to point **past** the node you are removing.  
+
+![remove a node #1](images/remove-node-1.png)
+![remove a node #2](images/remove-node-2.png)
+![remove a node #3](images/remove-node-3.png)
+![remove a node #4](images/remove-node-4.png)
+
+```python
+def remove(self, index):
+    if not self.head:
+        return
+
+    if index == 0:
+        self.head = self.head.next
+
+    # Traverse the list until you find the node at index -1
```

Comment:
This line was a little confusing to me at fist. I understand it to mean:

> Traverse the list until you find the node at the given index minus one

so if you're looking to remove index 4, you find the node at index 3 and operate from there. Does that sound about right?